addons/website/controllers/main.py

```diff
@@ -870,6 +902,56 @@ def theme_customize_bundle_reload(self):
     # Server actions
     # ------------------------------------------------------
 
+    @http.route(['/website/theme_upload_font'], type='json', auth='user', website=True)
+    def theme_upload_font(self, name, data):
+        """
+        Uploads font binary data and returns metadata about accessing individual fonts.
+        :param name: name of the uploaded file
+        :param data: binary content of the uploaded file
+        :return: list of dict describing each contained font with:
+            - name
+            - mimetype
+            - attachment id
+            - attachment URL
+        """
+        def create_attachment(font, data):
+            """ Creates font attachments right away to avoid keeping
+            several extracted contents in memory. """
+            ext = font['name'].split('.')[-1].lower()
+            font['mimetype'] = f'font/{ext}'
+            attachment = request.env['ir.attachment'].create({
+                'name': font['name'],
+                'mimetype': font['mimetype'],
+                'raw': data,
+                'public': True,
+            })
+            font['id'] = attachment.id
+            font['url'] = f"/web/content/{attachment.id}/{font['name']}"
+            return font
+
+        result = []
+        binary_data = base64.b64decode(data, validate=True)
+        readable_data = BytesIO(binary_data)
+        if zipfile.is_zipfile(readable_data):
+            with zipfile.ZipFile(readable_data, "r") as zip_file:
+                for entry in zip_file.filelist:
+                    if entry.file_size > MAX_FONT_FILE_SIZE:
+                        raise UserError(_("File '%s' exceeds maximum allowed file size", entry.filename))
+                    if entry.filename.rsplit('.', 1)[-1].lower() not in SUPPORTED_FONT_EXTENSIONS \
+                            or entry.filename.startswith('__MACOSX') \
+                            or '/.' in entry.filename:
```

Comment:
- The first condition just makes sure to only try to import font files based on their extension.
- If a zip is created on a Mac, a folder `__MACOSX` is included at the root of the zip to contain additional metadata - but the filenames inside that folder are the exact same one as in the actual data file hierarchy - therefore they would match the extension check of the first condition. The second condition exludes that folder.
- The third condition excludes files that begin with `.`, since they typically are Unix-style hidden files.